### PR TITLE
[CARBONDATA-3688] Add compressor name in data file name

### PR DESCRIPTION
**Pull request by jackylk**
### Why is this PR needed?
Currently there is no easy way to tell what compressor is used given one carbondata file unless to read the FileFooter in the file.
 
 ### What changes were proposed in this PR?
Added compressor name in the carbondata file name, now the file name pattern is 
partNo-taskNo-batchNo-bucketNo-segmentNo-timestamp.compressorName.carbondata

    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/165/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1868/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/171/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1873/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/172/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1874/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/176/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1879/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/177/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1880/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/184/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1887/

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java [hunk 1]
```diff
@@ -285,17 +286,39 @@ public static String getSegmentPath(String tablePath, String segmentId) {
...
+
+    // Start from CarbonData 2.0, the data file name patten is:
+    // partNo-taskNo-batchNo-bucketNo-segmentNo-timestamp.compressor.carbondata
+    // For example:
+    // part-0-0_batchno0-0-0-1580982686749.zstd.carbondata
+    //
+    // If the compressor name is missing, the file is compressed by snappy, which is
+    // the default compressor in CarbonData 1.x
+
+    return new StringBuffer().append(DATA_PART_PREFIX)
```
Comment (niuge01) [inline core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: There is need not use StringBuffer to build string, just use string concat will be ok.

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -1083,7 +1083,7 @@ private CarbonCommonConstants() {
    * The optional values are 'SNAPPY','GZIP','BZIP2','LZ4','ZSTD' and empty.
    * Specially, empty means that Carbondata will not compress the sort temp files.
    */
-  public static final String CARBON_SORT_TEMP_COMPRESSOR_DEFAULT = "SNAPPY";
+  public static final String CARBON_SORT_TEMP_COMPRESSOR_DEFAULT = "zstd";
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java [hunk 3]
```diff
@@ -163,7 +163,7 @@ public SegmentRefreshInfo getCommittedSegmentRefreshInfo(Segment segment, Update
     return segmentRefreshInfo;
   }
 
-  private String getSegmentID(String carbonIndexFileName, String indexFilePath) {
+  private String getTimestamp(String carbonIndexFileName, String indexFilePath) {
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/readcommitter/LatestFilesReadCommittedScope.java]: I changed back

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java (hunk 1 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: I changed to StringBuilder, and this link (https://stackoverflow.com/questions/47605/string-concatenation-concat-vs-operator) suggest StringBuilder is more efficient

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/193/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1895/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/200/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1902/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/203/

**Comment by QiangCai** [comment]
better to do further testing on a cluster, not only a local machine.

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1905/

**Comment by jackylk** [comment]
@QiangCai ok, I will verify on cluster

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java [hunk 4]
```diff
@@ -285,17 +286,38 @@ public static String getSegmentPath(String tablePath, String segmentId) {
...
+        .append("-")
+        .append(taskNo)
+        .append(BATCH_PREFIX)
+        .append(batchNo)
+        .append("-")
+        .append(bucketNumber)
+        .append("-")
+        .append(segmentNo)
+        .append("-")
+        .append(factUpdateTimeStamp)
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: supposed to append compressor name here ? I see that argument `compressor` is not used.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datacompaction/TableLevelCompactionOptionTest.scala [hunk 5]
```diff
@@ -139,6 +139,8 @@ class TableLevelCompactionOptionTest extends QueryTest
       sql(s"LOAD DATA LOCAL INPATH '$tempFilePath' INTO TABLE carbon_table")
     }
 
+    sql("SHOW SEGMENTS FOR TABLE carbon_table").show
```
Comment (ajantha-bhat) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datacompaction/TableLevelCompactionOptionTest.scala]: revert this or add validation

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala [hunk 6]
```diff
@@ -141,11 +141,6 @@ private[sql] case class CarbonDescribeFormattedCommand(
         Strings.formatSize(
           tblProps.getOrElse(CarbonCommonConstants.CARBON_LOAD_MIN_SIZE_INMB,
             CarbonCommonConstants.CARBON_LOAD_MIN_SIZE_INMB_DEFAULT).toFloat), ""),
-      ("Data File Compressor ", tblProps
```
Comment (ajantha-bhat) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala]: why removed this ?

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java [hunk 7]
```diff
@@ -188,7 +190,7 @@ private void initializeAtFirstRow() throws IOException {
       compressorName = carbonTable.getTableInfo().getFactTable().getTableProperties().get(
           CarbonCommonConstants.COMPRESSOR);
       if (null == compressorName) {
-        compressorName = CompressorFactory.getInstance().getCompressor().getName();
+        compressorName = CompressorFactory.NativeSupportedCompressor.SNAPPY.getName();
```
Comment (ajantha-bhat) [inline streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: why always snappy for streaming segment ? before I think it was supporting configured compressor

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java [hunk 8]
```diff
@@ -140,7 +140,9 @@ private void initialize(TaskAttemptContext job) throws IOException {
 
     segmentDir = CarbonTablePath.getSegmentPath(
         carbonTable.getAbsoluteTableIdentifier().getTablePath(), segmentId);
-    fileName = CarbonTablePath.getCarbonDataFileName(0, taskNo + "", 0, 0, "0", segmentId);
+    fileName = CarbonTablePath.getCarbonDataFileName(
```
Comment (ajantha-bhat) [inline streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: why always snappy for streaming segment ? before I think it was supporting configured compressor

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java (hunk 4 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: yes, I wanted to test once without changing the file name, in order to simulate using new jar to test on old files.
As the CI is passed, so that the compatibility is ok.
I will add the compressor name in file name now

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datacompaction/TableLevelCompactionOptionTest.scala (hunk 5 above)
Comment (jackylk) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/datacompaction/TableLevelCompactionOptionTest.scala]: fixed

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala (hunk 6 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala]: Because the default compressor changed, so this print is not correct now. And there is no way to know the compressor unless you check the file name. I think this is better since very loading can use different compressor, so we should not show in DESC which is table level

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java (hunk 7 above)
Comment (jackylk) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: There is testcase failing if set to ZSTD. Need another PR for it

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java (hunk 8 above)
Comment (jackylk) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: There is testcase failing if set to ZSTD. Need another PR for it

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/216/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/219/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1921/

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java (hunk 7 above)
Comment (ajantha-bhat) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: ok, please add a TODO

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java (hunk 8 above)
Comment (ajantha-bhat) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: ok, please add a TODO

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/257/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1960/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/386/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2087/

File: core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java [hunk 9]
```diff
@@ -409,11 +433,10 @@ public static String getTimeStampFromFileName(String carbonDataFileName) {
      * @return
      */
     public static Boolean compareCarbonFileTimeStamp(String fileName, Long timestamp) {
-      int lastIndexOfHyphen = fileName.lastIndexOf("-");
-      int lastIndexOfDot = fileName.lastIndexOf(".");
-      if (lastIndexOfHyphen > 0 && lastIndexOfDot > 0) {
-        return fileName.substring(fileName.lastIndexOf("-") + 1, fileName.lastIndexOf("."))
-            .equals(timestamp.toString());
+      int startIndex = fileName.lastIndexOf("-");
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: Suggest to use CarbonCommonConstants for "-". "."

Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/path/CarbonTablePath.java]: This func is not used anywhere, I deleted it

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/399/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2100/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/describeTable/TestDescribeTable.scala [hunk 10]
```diff
@@ -68,15 +68,6 @@ class TestDescribeTable extends QueryTest with BeforeAndAfterAll {
     assert(descPar.exists(_.toString().contains("Partition Parameters:")))
   }
 
-  test(testName = "Compressor Type update from carbon properties") {
-    sql("drop table if exists b")
-    sql(sqlText = "create table b(a int,b string) STORED AS carbondata")
-    CarbonProperties.getInstance().addProperty(CarbonCommonConstants.COMPRESSOR, "gzip")
-    val result = sql(sqlText = "desc formatted b").collect()
-    assert(result.filter(row => row.getString(0).contains("Data File Compressor")).head.getString
```
Comment (ajantha-bhat) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/describeTable/TestDescribeTable.scala]: As now we are not changing the default compressor, we can revert this change ?

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala [hunk 11]
```diff
@@ -141,11 +141,6 @@ private[sql] case class CarbonDescribeFormattedCommand(
...
-        .getOrElse(CarbonCommonConstants.COMPRESSOR,
-          CarbonProperties.getInstance()
```
Comment (ajantha-bhat) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala]: As now we are not changing the default compressor, we can revert this change ?

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java [hunk 12]
```diff
@@ -139,7 +139,9 @@ private void initialize(TaskAttemptContext job) throws IOException {
...
+        0, taskNo + "", 0, 0, "0",
+        segmentId, CompressorFactory.NativeSupportedCompressor.SNAPPY.getName());
```
Comment (ajantha-bhat) [inline streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: revert this or use CCC default compressor

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java [hunk 13]
```diff
@@ -186,7 +188,7 @@ private void initializeAtFirstRow() throws IOException {
       compressorName = carbonTable.getTableInfo().getFactTable().getTableProperties().get(
           CarbonCommonConstants.COMPRESSOR);
       if (null == compressorName) {
-        compressorName = CompressorFactory.getInstance().getCompressor().getName();
+        compressorName = CompressorFactory.NativeSupportedCompressor.SNAPPY.getName();
```
Comment (ajantha-bhat) [inline streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: revert this or use CCC default compressor

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/describeTable/TestDescribeTable.scala (hunk 10 above)
Comment (jackylk) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/describeTable/TestDescribeTable.scala]: No, this print is not correct. For example, when loading, user set to use ZSTD compressor, and then set to SNAPPY again after loading, then this command still prints SNAPPY which is misleading

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala (hunk 11 above)
Comment (jackylk) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/table/CarbonDescribeFormattedCommand.scala]: explained above

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java (hunk 13 above)
Comment (jackylk) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: fixed

File: streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java (hunk 12 above)
Comment (jackylk) [reply streaming/src/main/java/org/apache/carbondata/streaming/CarbonStreamRecordWriter.java]: fixed

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.4/419/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/2119/

**Comment by ajantha-bhat** [comment]
LGTM
